pkg/sentry/kernel/ipc/registry.go

```diff
@@ -0,0 +1,113 @@
+// Copyright 2021 The gVisor Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package ipc
+
+import (
+	"gvisor.dev/gvisor/pkg/abi/linux"
+	"gvisor.dev/gvisor/pkg/context"
+	"gvisor.dev/gvisor/pkg/log"
+	"gvisor.dev/gvisor/pkg/sentry/fs"
+	"gvisor.dev/gvisor/pkg/sentry/kernel/auth"
+	"gvisor.dev/gvisor/pkg/syserror"
+)
+
+// Registry is similar to Object, but for registries. It represent an abstract
+// SysV IPC registry with fields common to all SysV registries. Registry should
+// be included in registry implementattion of other SysV IPC mechanisms and
+// should be protected using a mutex.
+type Registry struct {
+	// UserNS owning the IPC namespace this registry belongs to. Immutable.
+	UserNS *auth.UserNamespace
+
+	// Objects is a map of IDs to IPC mechanisms.
+	Objects map[ID]Mechanism
```

Comment:
The three fields below probably shouldn't be public. 

Generally speaking, mutable fields of a type shouldn't be made public per the principle of [encapsulation](https://en.wikipedia.org/wiki/Encapsulation_(computer_programming)). Users of this object should interact with it through well-defined public methods, instead requiring knowledge about how to manipulate its internal state.

For example, to register a new object, a user has to read and update `LastIDUsed` and `KeysToIDs`. What if a future change requires you to update a new field? As currently implemented, you'd have to change all users of this object, and failing to update one would lead to a subtle bug. If you define a `Register` method instead, the logic for registering an object would be defined in one location and the internal state is kept private to the registry.

Reply:
Done. All mutable fields can now be only modified using operation-specific methods. Only `UserNS` is exposed.

There are two cases, [here](https://github.com/sudo-sturbia/gvisor/blob/8baef691734f2e03b9862b94b78e1d7fc08c2edf/pkg/sentry/kernel/shm/shm.go#L287), and [here](https://github.com/sudo-sturbia/gvisor/blob/8baef691734f2e03b9862b94b78e1d7fc08c2edf/pkg/sentry/kernel/semaphore/semaphore.go#L293), where all objects are retrieved. You'll notice that in both cases, `objects` is only read from, in fact, even the IDs are not needed. I created `Registry.Objects` which returns the objects map. While it works, it doesn't look great, because it doesn't necessarily limit the caller to only reading. One thing that can be done is to make `Objects` return a newly allocated slice of mechanisms, which is easily done, but adds somewhat unnecessary overhead. I'm curious what you think?